### DISCUSSION TITLE
Ideas for API improvements

### DISCUSSION THREAD

**Discussion by titaniumtraveler** [winnow-rs/winnow#392] [Ideas]
I found a few instances where the winnow API could be nicer and instead of being slightly annoyed about it not being that, I thought I would rather document them here and see what other people think about these ideas.

1. `ParserIterator` should implement iterator on `Self` instead of `&'a mut Self`:
  This point is pretty important for me as it makes the API just nicer
  and has no apparent drawbacks.
  Especially because `Iterator` should almost always be implemented on `Self`,
  because the standard library already contains `impl<T: Iterator> Iterator for &mut T`,
  making an impl on `&mut Self` kind of redunant.
2. Similar to `Iterator`, `Parser` should have an  `impl<T: Parser> Parser for &mut Parser`-impl as well.
  This can be useful when having a parser that contains some inner parser
  and enables just using the normal chaining functions on `&mut inner_parser` instead of needing to use `.parse_next()` manually.

Note that both changes *might* require major version bump,
though I'm not sure because of `1.` the same impl should then be covered by the `impl<T: Iterator> Iterator for &mut T` of iterator
(though theoretically someone could depend on `&mut &mut ParserIterator` implementing `Iterator`.)
and for `2.` it *should not* conflict with any existing code, but still might.

I will add to this list if I find more things, which could be improved.

**Comment by titaniumtraveler** [comment]
3. [`Parser::map`](https://docs.rs/winnow/latest/winnow/trait.Parser.html#method.map)
  should take `FnMut` instead of `Fn`, as it is much more useful.

**Reply by epage** [reply]
Looks like `nom`s `map` supported this for a while (rust-bakery/nom#1202) but not `Parser::map` until rust-bakery/nom#1476 and we missed pulling this in.

As `FnMut` is a wider bound, this should be a safe change.

**Reply by epage** [reply]
Released in v0.5.27

**Comment by comparenetworthbloilg** [comment]
is there any algorithm or program which detect when [dating a prisoner](https://needselfdevelopment.com/13-signs-an-inmate-really-loves-you/) is process should be done.

**Comment by epage** [comment]
> 2\. Similar to `Iterator`, `Parser` should have an  `impl<T: Parser> Parser for &mut Parser`-impl as well.
>         This can be useful when having a parser that contains some inner parser
>         and enables just using the normal chaining functions on `&mut inner_parser` instead of needing to use `.parse_next()` manually.

This requires a blanket `impl` which conflicts with our existing blanket `impl` for `FnMut`s which allows any function to be a parser implicitly.  We do offer `Parser::by_ref` to help with this.

**Comment by epage** [comment]
> 1. `ParserIterator` should implement iterator on `Self` instead of `&'a mut Self`:
>        This point is pretty important for me as it makes the API just nicer
>        and has no apparent drawbacks.
>        Especially because `Iterator` should almost always be implemented on `Self`,
>        because the standard library already contains `impl<T: Iterator> Iterator for &mut T`,
>        making an impl on `&mut Self` kind of redunant.

This was initially reported against nom in rust-bakery/nom#1586 and then fixed in nom-bakery/nom#1587, after we forked.

I didn't initially fix it or copy over the fix due to lack of experience with `iterator` making it difficult for me to evaluate.

For example, `iterator` doesn't report errors in-line but requires you to access the returned value after iteration is complete.  By supporting `Iterator` on `ParserIterator`, it will be even easier for people to overlook this, potentially leading to bugs.

**Reply by epage** [reply]
see also #349
